File: tugas_6.py
import csv

# CLASS NODE
class Node:
    def __init__(self, id, nama):
        self.id = id
        self.nama = nama
        self.height = 1
        self.left = None
        self.right = None


# CLASS AVL TREE
class AVL:
    def __init__(self):
        self.root = None

    # HEIGHT
    def height(self, node):
        return 0 if not node else node.height

    # BALANCE FACTOR
    def get_balance(self, node):
        return 0 if not node else self.height(node.left) - self.height(node.right)

    # RIGHT ROTATE
    def right_rotate(self, y):
        x = y.left
        t2 = x.right

        x.right = y
        y.left = t2

        y.height = 1 + max(self.height(y.left), self.height(y.right))
        x.height = 1 + max(self.height(x.left), self.height(x.right))

        return x

    # LEFT ROTATE
    def left_rotate(self, x):
        y = x.right
        t2 = y.left

        y.left = x
        x.right = t2

        x.height = 1 + max(self.height(x.left), self.height(x.right))
        y.height = 1 + max(self.height(y.left), self.height(y.right))

        return y

    # INSERT
    def insert(self, node, id, nama):
        if not node:
            return Node(id, nama)

        if id < node.id:
            node.left = self.insert(node.left, id, nama)
        elif id > node.id:
            node.right = self.insert(node.right, id, nama)
        else:
            return node

        node.height = 1 + max(self.height(node.left), self.height(node.right))

        balance = self.get_balance(node)

        # LL
        if balance > 1 and id < node.left.id:
            return self.right_rotate(node)

        # RR
        if balance < -1 and id > node.right.id:
            return self.left_rotate(node)

        # LR
        if balance > 1 and id > node.left.id:
            node.left = self.left_rotate(node.left)
            return self.right_rotate(node)

        # RL
        if balance < -1 and id < node.right.id:
            node.right = self.right_rotate(node.right)
            return self.left_rotate(node)

        return node

    # MIN VALUE
    def min_value(self, node):
        current = node
        while current.left:
            current = current.left
        return current

    # DELETE
    def delete(self, node, id):
        if not node:
            return node

        if id < node.id:
            node.left = self.delete(node.left, id)
        elif id > node.id:
            node.right = self.delete(node.right, id)
        else:
            if not node.left or not node.right:
                node = node.left if node.left else node.right
            else:
                temp = self.min_value(node.right)
                node.id = temp.id
                node.nama = temp.nama
                node.right = self.delete(node.right, temp.id)

        if not node:
            return node

        node.height = 1 + max(self.height(node.left), self.height(node.right))

        balance = self.get_balance(node)

        # LL
        if balance > 1 and self.get_balance(node.left) >= 0:
            return self.right_rotate(node)

        # LR
        if balance > 1 and self.get_balance(node.left) < 0:
            node.left = self.left_rotate(node.left)
            return self.right_rotate(node)

        # RR
        if balance < -1 and self.get_balance(node.right) <= 0:
            return self.left_rotate(node)

        # RL
        if balance < -1 and self.get_balance(node.right) > 0:
            node.right = self.right_rotate(node.right)
            return self.left_rotate(node)

        return node

    # SEARCH
    def search(self, node, id):
        if not node or node.id == id:
            return node

        if id < node.id:
            return self.search(node.left, id)

        return self.search(node.right, id)

    # TRAVERSAL
    def inorder(self, node):
        if node:
            self.inorder(node.left)
            print(f"ID: {node.id} - Nama: {node.nama}")
            self.inorder(node.right)

    def preorder(self, node):
        if node:
            print(f"ID: {node.id} - Nama: {node.nama}")
            self.preorder(node.left)
            self.preorder(node.right)

    def postorder(self, node):
        if node:
            self.postorder(node.left)
            self.postorder(node.right)
            print(f"ID: {node.id} - Nama: {node.nama}")

    # IMPORT CSV
    def import_csv(self, path):
        try:
            with open(path, newline='') as file:
                reader = csv.reader(file)
                next(reader)  # skip header

                for row in reader:
                    if len(row) >= 2:
                        try:
                            id = int(row[0])
                            nama = row[1]
                            self.root = self.insert(self.root, id, nama)
                        except ValueError:
                            print("Data salah:", row)

            print("Import CSV berhasil!")

        except Exception as e:
            print("Error:", e)


# MAIN PROGRAM
def main():
    tree = AVL()

    while True:
        print("\n=== MENU AVL TREE ===")
        print("1. Tambah Data")
        print("2. Cari Data")
        print("3. Hapus Data")
        print("4. Inorder")
        print("5. Preorder")
        print("6. Postorder")
        print("7. Import CSV")
        print("8. Keluar")

        pilihan = input("Pilih: ")

        if pilihan == "1":
            id = int(input("ID: "))
            nama = input("Nama: ")
            tree.root = tree.insert(tree.root, id, nama)

        elif pilihan == "2":
            cari = int(input("Cari ID: "))
            hasil = tree.search(tree.root, cari)

            if hasil:
                print("Ditemukan:", hasil.nama)
            else:
                print("Tidak ditemukan")

        elif pilihan == "3":
            hapus = int(input("Hapus ID: "))
            tree.root = tree.delete(tree.root, hapus)
            print("Data dihapus")

        elif pilihan == "4":
            tree.inorder(tree.root)

        elif pilihan == "5":
            tree.preorder(tree.root)

        elif pilihan == "6":
            tree.postorder(tree.root)

        elif pilihan == "7":
            path = input("Path CSV: ")
            tree.import_csv(path)

        elif pilihan == "8":
            print("Program selesai")
            break

        else:
            print("Pilihan tidak valid")


if __name__ == "__main__":
    main()

File: tugas_6.java
import java.util.Scanner;
import java.io.BufferedReader;
import java.io.FileReader;

// CLASS NODE AVL
class Node {
    int id;
    String nama;
    int height;
    Node left, right;

    Node(int id, String nama) {
        this.id = id;
        this.nama = nama;
        this.height = 1;
        left = right = null;
    }
}

// CLASS AVL TREE
class AVL {
    Node root;

    // HEIGHT
    int height(Node n) {
        return (n == null) ? 0 : n.height;
    }

    // BALANCE FACTOR
    // -1, 0, atau 1 maka seimbang
    int getBalance(Node n) {
        return (n == null) ? 0 : height(n.left) - height(n.right);
    }

    // RIGHT ROTATE
    Node rightRotate(Node y) {
        Node x = y.left; // ambil anak kiri dan jadikan node
        Node T2 = x.right; // simpan subtree kanan dari x

        x.right = y; // jadikan y sebagai anak kanan
        y.left = T2; 

        y.height = Math.max(height(y.left), height(y.right)) + 1;
        x.height = Math.max(height(x.left), height(x.right)) + 1;

        return x;
    }

    // LEFT ROTATE             // contoh
    Node leftRotate(Node x) { // x = 10
        Node y = x.right;  // y = 20
        Node T2 = y.left; // T2 = null

        y.left = x;
        x.right = T2;

        x.height = Math.max(height(x.left), height(x.right)) + 1;
        y.height = Math.max(height(y.left), height(y.right)) + 1;

        return y;
    }

    // INSERT AVL
    Node insert(Node node, int id, String nama) {
        if (node == null)
            return new Node(id, nama);

        if (id < node.id)
            node.left = insert(node.left, id, nama);
        else if (id > node.id)
            node.right = insert(node.right, id, nama);
        else
            return node;

        // update height node
        node.height = 1 + Math.max(height(node.left), height(node.right));

        //hitung balance faktor
        int balance = getBalance(node);

        // Left berat
        if (balance > 1 && id < node.left.id)
            return rightRotate(node);

        // kanan berat
        if (balance < -1 && id > node.right.id)
            return leftRotate(node);

        // kombinasi kedua
        if (balance > 1 && id > node.left.id) {
            node.left = leftRotate(node.left); // rotasi kiri dulu baru kanan
            return rightRotate(node);
        }

        // RL
        if (balance < -1 && id < node.right.id) {
            node.right = rightRotate(node.right);
            return leftRotate(node);
        }

        return node;
    }

    // MIN VALUE
    Node minValue(Node node) {
        Node current = node;
        while (current.left != null)
            current = current.left;
        return current;
    }

    // DELETE AVL
    Node delete(Node root, int id) {
        if (root == null)
            return root;

        if (id < root.id)
            root.left = delete(root.left, id);
        else if (id > root.id)
            root.right = delete(root.right, id);
        else {
            if ((root.left == null) || (root.right == null)) {
                Node temp = (root.left != null) ? root.left : root.right;

                if (temp == null) {
                    root = null;
                } else {
                    root = temp;
                }
            } else {
                Node temp = minValue(root.right);

                root.id = temp.id;
                root.nama = temp.nama;

                root.right = delete(root.right, temp.id);
            }
        }

        if (root == null)
            return root;

        root.height = 1 + Math.max(height(root.left), height(root.right));

        int balance = getBalance(root);

        // LL
        if (balance > 1 && getBalance(root.left) >= 0)
            return rightRotate(root);

        // LR
        if (balance > 1 && getBalance(root.left) < 0) {
            root.left = leftRotate(root.left);
            return rightRotate(root);
        }

        // RR
        if (balance < -1 && getBalance(root.right) <= 0)
            return leftRotate(root);

        // RL
        if (balance < -1 && getBalance(root.right) > 0) {
            root.right = rightRotate(root.right);
            return leftRotate(root);
        }

        return root;
    }

    // SEARCH
    Node search(Node root, int id) {
        if (root == null || root.id == id)
            return root;

        if (id < root.id)
            return search(root.left, id);

        return search(root.right, id);
    }

    // TRAVERSAL
    void inorder(Node root) {
        if (root != null) {
            inorder(root.left);
            System.out.println("ID: " + root.id + " - Nama: " + root.nama);
            inorder(root.right);
        }
    }

    void preorder(Node root) {
        if (root != null) {
            System.out.println("ID: " + root.id + " - Nama: " + root.nama);
            preorder(root.left);
            preorder(root.right);
        }
    }

    void postorder(Node root) {
        if (root != null) {
            postorder(root.left);
            postorder(root.right);
            System.out.println("ID: " + root.id + " - Nama: " + root.nama);
        }
    }

    // IMPORT CSV
    void importCSV(String path) {
        try {
            BufferedReader br = new BufferedReader(new FileReader(path));
            String line;

            br.readLine(); // skip header

            while ((line = br.readLine()) != null) {
                String[] data = line.split(",");

                if (data.length >= 2) {
                    try {
                        int id = Integer.parseInt(data[0]);
                        String nama = data[1];

                        root = insert(root, id, nama);

                    } catch (NumberFormatException e) {
                        System.out.println("Data salah: " + line);
                    }
                }
            }

            br.close();
            System.out.println("Import CSV berhasil!");

        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }
    }
}

// MAIN CLASS
public class tugas_6 {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        AVL tree = new AVL();
        int pilihan;

        do {
            System.out.println("\n=== MENU AVL TREE ===");
            System.out.println("1. Tambah Data");
            System.out.println("2. Cari Data");
            System.out.println("3. Hapus Data");
            System.out.println("4. Inorder");
            System.out.println("5. Preorder");
            System.out.println("6. Postorder");
            System.out.println("7. Import CSV");
            System.out.println("8. Keluar");
            System.out.print("Pilih: ");
            pilihan = input.nextInt();

            switch (pilihan) {
                case 1:
                    System.out.print("ID: ");
                    int id = input.nextInt();
                    input.nextLine();

                    System.out.print("Nama: ");
                    String nama = input.nextLine();

                    tree.root = tree.insert(tree.root, id, nama);
                    break;

                case 2:
                    System.out.print("Cari ID: ");
                    int cari = input.nextInt();

                    Node hasil = tree.search(tree.root, cari);

                    if (hasil != null)
                        System.out.println("Ditemukan: " + hasil.nama);
                    else
                        System.out.println("Tidak ditemukan");
                    break;

                case 3:
                    System.out.print("Hapus ID: ");
                    int hapus = input.nextInt();

                    tree.root = tree.delete(tree.root, hapus);
                    System.out.println("Data dihapus");
                    break;

                case 4:
                    tree.inorder(tree.root);
                    break;

                case 5:
                    tree.preorder(tree.root);
                    break;

                case 6:
                    tree.postorder(tree.root);
                    break;

                case 7:
                    input.nextLine();
                    System.out.print("Path CSV: ");
                    String path = input.nextLine();

                    tree.importCSV(path);
                    break;

                case 8:
                    System.out.println("Program selesai");
                    break;
            }

        } while (pilihan != 8);

        input.close();
    }
}